src/python/gudhi/_pers_cub_low_dim.cc

```diff
@@ -5,78 +5,87 @@
  *    Copyright (C) 2023 Inria
  *
  *    Modification(s):
+ *      - 2025/01 Vincent Rouvreau: Use nanobind instead of PyBind11 for python bindings
  *      - YYYY/MM Author: Description of the modification
  */
 
 #include <vector>
-#include <array>
 #include <limits>
-#include <stdexcept>
 
-#include <pybind11/pybind11.h>
-#include <pybind11/stl.h>
-#include <pybind11/stl_bind.h>
-#include <pybind11/numpy.h>
+#include <nanobind/nanobind.h>
+#include <nanobind/ndarray.h>
 
 #include <boost/range/counting_range.hpp>
 #include <boost/range/adaptor/transformed.hpp>
 
 #include <gudhi/Persistence_on_a_line.h>
 #include <gudhi/Persistence_on_rectangle.h>
-#include <gudhi/Debug_utils.h>
 
-namespace py = pybind11;
-typedef std::vector<std::array< float, 2>> Vf;
-typedef std::vector<std::array<double, 2>> Vd;
-PYBIND11_MAKE_OPAQUE(Vf);
-PYBIND11_MAKE_OPAQUE(Vd);
+namespace nb = nanobind;
 
-template<class T>
-py::array wrap_persistence_1d(py::array_t<T> data) {
-  py::buffer_info buf = data.request();
-  if(buf.ndim!=1)
-    throw std::runtime_error("Data must be a 1-dimensional array");
-  auto cnt = boost::counting_range<py::ssize_t>(0, buf.shape[0]);
-  char* p = static_cast<char*>(buf.ptr);
-  auto h = buf.strides[0];
-  auto proj = [=](py::ssize_t i){ return *reinterpret_cast<T*>(p + i * h); };
+template <class T>
+auto wrap_persistence_1d(nb::ndarray<T, nb::ndim<1>> data)
+{
+  auto cnt = boost::counting_range<nb::ssize_t>(0, data.shape(0));
+  auto proj = [&](nb::ssize_t i) { return data(i); };
```

Comment:
Should we create a view first, and then index that?
https://nanobind.readthedocs.io/en/latest/ndarray.html#fast-array-views
(no idea if it is useful for this case)

Reply:
Tested it with the same benchmark as above and the difference is significant! I gained 10 seconds on the tests with 10000000 points. So, I would definitely keep it and add it on other bindings too.